src/commands/test/runner.py

```diff
@@ -196,29 +184,66 @@ async def invoke_test_function(self, function_name: str):
-                    f"Expected an exception matching the following error: {self._expected_error}"
-                )
-            return call_result
-
-        except StarkException as ex:
-            if self._expected_error:
-                if not self._expected_error.match(ex):
-                    raise ExceptMismatchException(
-                        expected_name=self._expected_error.name,
-                        expected_message=self._expected_error.message,
-                        received=ex,
+            try:
+                try:
+                    call_result = await func().invoke()
+                    for listener in self._test_finished_listener_map.values():
+                        if listener:
+                            listener()
+
+                    if self._expected_error is not None:
+                        raise StandardReportedException(
+                            f"Expected an exception matching the following error:\n{self._expected_error}"
+                        )
+                    return call_result
+                # catch StarkNet exceptions and wrap them with RevertableException
+                except StarkException as ex:
+                    raise RevertableException(
+                        error_type=ex.code.name,
+                        error_message=extract_core_info_from_stark_ex_message(
+                            ex.message
+                        ),
+                        exception=ex,
                     ) from ex
-            else:
-                raise StarkReportedException(ex) from ex
+            # catch a RevertableException and rethrow the original_exception if the exception is not expected
+            except RevertableException as ex:
+                if self._expected_error:
+                    if not self._expected_error.match(ex):
+                        raise ExpectedRevertMismatchException(
+                            expected=self._expected_error,
+                            received=ex,
+                        ) from ex
+                else:
+                    if ex.original_exception:
+                        raise ex.original_exception
+        # wrap exceptions with StandardReportedException print errors in a consistent way
+        except StarkException as ex:
+            raise StandardReportedException(
+                error_message=extract_core_info_from_stark_ex_message(ex.message),
+                error_type=ex.code.name,
+                code=ex.code.value,
+                details=ex.message,
+            ) from ex
+        except BaseException as ex:
+            print(ex)
+            raise ex
         finally:
             CairoFunctionRunner.run_from_entrypoint = original_run_from_entrypoint
             self._expected_error = None
 
+    def subscribe_to_test_finish(
+        self, listener: Callable[[], None]
+    ) -> Callable[[], None]:
+        listener_id = uuid()
+        self._test_finished_listener_map[listener_id] = listener
```

Comment:
If we remove ` stop_expecting_emit` we can use a `set()` here

Reply:
I would also name it `add_test_finish_hook` subscribe suggests some signaler

Reply:
Aside, I really like flexibility of hook-based approach, good job

Reply:
`stop_expecting_emit` isn't returned by the cheatcode. It is only called by the "test finished hook", so I have to preserve this function (under new name).